### PR TITLE
fix buggy handling of partial ingress packet data

### LINKED ISSUE #246: python/meterpreter/reverse_tcp_ssl appears broken
Automated testing is failing on python/meterpreter/reverse_tcp_ssl.  A session is created, but it looks like the stdapi stuff does not get loaded:
```
> use exploit/multi/handler
> set payload python/meterpreter/reverse_tcp_ssl
> set LHOST 192.168.135.112
LHOST => 192.168.135.112
> set LPORT 30001
LPORT => 30001
> show options

Module options (exploit/multi/handler):

   Name  Current Setting  Required  Description
   ----  ---------------  --------  -----------


Payload options (python/meterpreter/reverse_tcp_ssl):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST  192.168.135.112  yes       The listen address
   LPORT  30001            yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Wildcard Target


> run -z
[*] Exploit running as background job 0.
[*] Started reverse SSL handler on 192.168.135.112:30001 
[*] Sending stage (42231 bytes) to 192.168.134.133

> sessions -C sysinfo
[*] Running 'sysinfo' on meterpreter session 1 (192.168.134.133)
[-] Unknown command: sysinfo.
> sessions -C ifconfig
[*] Running 'ifconfig' on meterpreter session 1 (192.168.134.133)
[-] Unknown command: ifconfig.
> sessions -C sessions -l

Active sessions
===============

  Id  Name  Type                       Information  Connection
  --  ----  ----                       -----------  ----------
  1         meterpreter python/python               192.168.135.112:30001 -> 192.168.134.133:49178 (192.168.134.133)

> sessions -C getuid
[*] Running 'getuid' on meterpreter session 1 (192.168.134.133)
[-] Unknown command: getuid.
> loadpath test/modules
Loaded 35 modules:
    9 posts
    13 exploits
    13 auxiliarys
> use post/test/meterpreter
> set verbose true
verbose => true
> set addentropy true
addentropy => true
> set session 1
session => 1
> run
[-] Post failed: NoMethodError undefined method `dir' for nil:NilClass
[-] Call stack:
[-]   /home/msfuser/rapid7/metasploit-framework/test/modules/post/test/meterpreter.rb:37:in `setup'
[*] Cleanup: changing working directory back to 
[-] Post failed: NoMethodError undefined method `dir' for nil:NilClass
[-] Call stack:
[-]   /home/msfuser/rapid7/metasploit-framework/test/modules/post/test/meterpreter.rb:373:in `cleanup'
> use post/test/railgun
> set verbose true
verbose => true
> set session 1
session => 1
> run
[!] SESSION may not be compatible with this module.
[-] Post failed: RuntimeError Could not get a hold of the session.
[-] Call stack:
[-]   /home/msfuser/rapid7/metasploit-framework/lib/msf/core/post_mixin.rb:63:in `check_for_session_readiness'
[-]   /home/msfuser/rapid7/metasploit-framework/lib/msf/core/post_mixin.rb:45:in `setup'
[-]   /home/msfuser/rapid7/metasploit-framework/lib/msf/core/post.rb:38:in `setup'
[*] Post module execution completed
> exit -y

```

My testing was on python 2.7 x64 and windows, though I expect this will be universal:

image

Is there a chance the recent changes to tlv encryption did this?  I've kicked off tests for x86 python 2.7, but I anticipate the same result.

### PR DESCRIPTION
**Pull request by busterb**
If we have more data, and the packet parser needs more data, connect the two together rather than bailing. This fixes https://github.com/rapid7/metasploit-payloads/issues/246 reverse_tcp_ssl along with probably a lot of other higher-latency corner cases when a payload doesn't seem to stage stdapi or other initial components.

## Verification

Passes all of the payload tests, namely @bwatters-r7 test case for reverse_tcp_ssl

### CODE REVIEW

File: lib/rex/post/meterpreter/packet_parser.rb [hunk 1]
```diff
@@ -30,24 +30,20 @@ def reset
   # Reads data from the wire and parse as much of the packet as possible.
```
Comment (busterb) [inline lib/rex/post/meterpreter/packet_parser.rb]: Note, now this statement is no longer a lie.

**Comment by bwatters-r7** [comment]
image

Clone 3 is a little buggy.  This looks great!

**Comment by busterb** [comment]
Since this is in the middle of core code, would you mind running a wider payload shakedown as well?

**Comment by bwatters-r7** [comment]
Sure; would 
image

That first error is a really troublesome bug that I think exists in the vmware networking?  It is a super long story.  Ignore it.

**Comment by busterb** [comment]
Add an HTTP check and I think we have a winner.

**Comment by bwatters-r7** [comment]
Already started the http, but I was going to run a  couple stageless, python, and an RC4.

**Comment by bwatters-r7** [comment]
image

**Comment by busterb** [comment]
LGTM - think it'll be done before the snapshot occurs in 10 minutes?

**Review by sempervictus — COMMENTED** [review]
Pulling this in, you may have nailed a lot of birds with this stone. Awesome find, thank you.

**Comment by OJ** [comment]
Interesting. Prior to me modifying the code in this function I didn't think there was a loop that read in the entire header, which is why I didn't add one in. In hindsight, this does seem like a bad idea. Interesting that prior to the changes we didn't have any issues (nor did I see any when I tested).

@bcook-r7  I'll have a play with it at some point today. Can you give me a scenario where things break prior to this change? Thank you!

**Comment by busterb** [comment]
Sure! The case we were seeing failing reliably was using php/meterpreter_reverse_tcp_ssl

**Comment by bwatters-r7** [comment]
I ran this through another three payload types with the same results.  Everything passed.

**Comment by bwatters-r7** [comment]
# Release notes
This patches a bug found in our initial handshake with meterpreter that caused some payloads (namely python/meterpreter/reverse_tcp_ssl to fail in establishing a clean session.

**Comment by bwatters-r7** [comment]
Fixes https://github.com/rapid7/metasploit-payloads/issues/246

**Comment by sempervictus** [comment]
Agreed - works here.
Why did that payload work for years before though? We need an erd-like flow diagram of the transport tiers... Enough dark corners to hide a family of xenomorphs made from an elephant herd.
